File: Source/Init/Difficulty/Difficulty.cs
using System;
using WCSharp.Api;

public static class Difficulty
{
    public static int DifficultyValue { get; private set; }
    public static DifficultyOption DifficultyOption { get; private set; }
    public static bool IsDifficultyChosen { get; set; } = false;

    private const float TimeBeforeShowingVote = 2.0f;
    private const float TimeToChooseDifficulty = 10.0f;

    private static SelectionDialog voteDialog;

    public static void Initialize()
    {
        try
        {
            if (Gamemode.CurrentGameMode != GameMode.Standard) return;
            if (IsDifficultyChosen) return;

            DifficultyOption.Initialize();
            voteDialog = BuildVoteDialog();

            Utility.SimpleTimer(TimeBeforeShowingVote, ChooseDifficulty);
        }
        catch (Exception e)
        {
            Logger.Critical($"Error in Difficulty.Initialize: {e.Message}");
            throw;
        }
    }

    private static SelectionDialog BuildVoteDialog()
    {
        var dialog = new SelectionDialog(
            $"{Colors.COLOR_GOLD}Please choose a difficulty{Colors.COLOR_RESET}",
            closeOnSelect: false);

        foreach (var option in DifficultyOption.Options)
        {
            dialog.AddOption(option.ToString(), votingPlayer => RecordVote(option, votingPlayer));
        }

        return dialog;
    }

    private static void RecordVote(DifficultyOption option, player votingPlayer)
    {
        option.TallyCount++;
        Utility.TimedTextToAllPlayers(3.0f, $"{Colors.PlayerNameColored(votingPlayer)}|r has chosen {option} difficulty.{Colors.COLOR_RESET}");
    }

    private static void ChooseDifficulty()
    {
        voteDialog.ShowTo(Globals.ALL_PLAYERS);
        Utility.SimpleTimer(TimeToChooseDifficulty, TallyingVotes);
    }

    private static void TallyingVotes()
    {
        // Guards against the (already-scheduled) timer firing after the
        // difficulty was already forced through some other path, e.g.
        // ChangeDifficulty being called by an admin mid-vote.
        if (IsDifficultyChosen) return;

        DifficultyOption pickedOption = null;
        var highestTallyCount = 0;

        foreach (var option in DifficultyOption.Options)
        {
            if (option.TallyCount > highestTallyCount)
            {
                highestTallyCount = option.TallyCount;
                pickedOption = option;
            }
        }

        // Nobody voted: fall back to Normal rather than leaving pickedOption
        // null (the previous version would null-ref here).
        SetDifficulty(pickedOption ?? FindOption(DifficultyLevel.Normal));
    }

    private static void SetDifficulty(DifficultyOption difficulty)
    {
        voteDialog?.Cancel();

        DifficultyOption = difficulty;
        DifficultyValue = difficulty.Value;
        IsDifficultyChosen = true;
        SetupGamemodeBasedOnDifficulty();
        Console.WriteLine($"{Colors.COLOR_YELLOW_ORANGE}The difficulty has been set to |r{difficulty}{Colors.COLOR_RESET}");
    }

    private static void SetupGamemodeBasedOnDifficulty()
    {
        Gamemode.NumberOfRounds = DifficultyValue == (int)DifficultyLevel.Progressive ? 3 : 5;
    }

    private static DifficultyOption FindOption(DifficultyLevel level)
    {
        return DifficultyOption.Options.Find(o => o.Value == (int)level);
    }

    /// <summary>
    /// Changes the difficulty of the game to the specified difficulty,
    /// bypassing the vote. Cancels any vote still in progress so players
    /// aren't left looking at a dialog whose result will never matter.
    /// </summary>
    /// <param name="difficulty">"normal", "hard", "impossible", ...</param>
    public static bool ChangeDifficulty(string difficulty = "normal")
    {
        difficulty = difficulty.ToLower();

        foreach (var option in DifficultyOption.Options)
        {
            var name = option.Name.ToLower();

            if (name.Contains(difficulty) || name.StartsWith(difficulty))
            {
                SetDifficulty(option);
                return true;
            }
        }
        return false;
    }
}

File: src/Init/Difficulty/Difficulty.ts
import { Logger } from 'src/Events/Logger/Logger'
import { CurrentGameMode } from 'src/Gamemodes/CurrentGameMode'
import { GameMode } from 'src/Gamemodes/GameModeEnum'
import { Globals } from 'src/Global/Globals'
import { Colors } from 'src/Utility/Colors/Colors'
import { ColorUtils } from 'src/Utility/Colors/ColorUtils'
import { Utility } from 'src/Utility/Utility'
import { getTriggerPlayer } from 'src/Utility/w3tsUtils'
import { Trigger } from 'w3ts'
import { DifficultyOption } from './DifficultyOption'

export class Difficulty {
    public static DifficultyValue: number
    public static DifficultyOption: DifficultyOption
    public static IsDifficultyChosen = false
    private static TIME_TO_CHOOSE_DIFFICULTY = 10.0
    private static triggerHandle: Trigger

    public static Initialize = () => {
        try {
            if (CurrentGameMode.active !== GameMode.Standard) return
            if (Difficulty.IsDifficultyChosen) return

            DifficultyOption.Initialize()
            DifficultyOption.DifficultyChoosing.setMessage(
                `${Colors.COLOR_GOLD}Please choose a difficulty${Colors.COLOR_RESET}`
            )
            Difficulty.RegisterSelectionEvent()

            Utility.SimpleTimer(2.0, () => Difficulty.ChooseDifficulty())
        } catch (e) {
            Logger.Critical(`Error in Difficulty.Initialize: ${e}`)
            throw e
        }
    }

    private static RegisterSelectionEvent = () => {
        Difficulty.triggerHandle ??= Trigger.create()!
        Difficulty.triggerHandle.registerDialogEvent(DifficultyOption.DifficultyChoosing)
        Difficulty.triggerHandle.addAction(() => {
            const player = getTriggerPlayer()
            const button = GetClickedButton()

            const option = DifficultyOption.Options.find(o => o.Button.handle === button)
            if (option) option.TallyCount++

            DifficultyOption.DifficultyChoosing.display(player, false)
            Utility.TimedTextToAllPlayers(
                3.0,
                `${ColorUtils.PlayerNameColored(player)}|r has chosen ${option?.toString()} difficulty.${Colors.COLOR_RESET}`
            )
        })
    }

    private static ChooseDifficulty = () => {
        for (const player of Globals.ALL_PLAYERS) DifficultyOption.DifficultyChoosing.display(player, true)
        Utility.SimpleTimer(Difficulty.TIME_TO_CHOOSE_DIFFICULTY, () => Difficulty.TallyingVotes())
    }

    private static TallyingVotes = () => {
        let highestTallyCount = 0
        let pickedOption: DifficultyOption | null = null

        for (const option of DifficultyOption.Options) {
            if (option.TallyCount > highestTallyCount) {
                highestTallyCount = option.TallyCount
                pickedOption = option
            }
        }

        Difficulty.RemoveDifficultyDialog()
        if (!pickedOption) print('picked option is null')
        if (pickedOption) Difficulty.SetDifficulty(pickedOption)
    }

    private static SetDifficulty = (difficulty: DifficultyOption) => {
        Difficulty.DifficultyOption = difficulty
        Difficulty.DifficultyValue = difficulty.Value
        Difficulty.IsDifficultyChosen = true
        print(
            `${Colors.COLOR_YELLOW_ORANGE}The difficulty has been set to |r${difficulty.toString()}${Colors.COLOR_RESET}`
        )
    }

    private static RemoveDifficultyDialog = () => {
        for (const player of Globals.ALL_PLAYERS) DifficultyOption.DifficultyChoosing.display(player, false)
    }

    /// <summary>
    /// Changes the difficulty of the game to the specified difficulty.
    /// </summary>
    /// <param name="difficulty">"normal", "hard", "impossible"</param>
    public static ChangeDifficulty = (difficulty = 'normal') => {
        for (let i = 0; i < DifficultyOption.Options.length; i++) {
            const option = DifficultyOption.Options[i]
            if (option.name.toLowerCase() === difficulty.toLowerCase()) {
                Difficulty.SetDifficulty(option)
                return true
            }
        }
        return false
    }
}
